### DISCUSSION TITLE
How add the lv2 path to the config.

### DISCUSSION THREAD

**Discussion by BorisSutin** [rerdavies/pipedal#179] [General]
Hello! Tell me how to add the lv2 path to the config.

**Comment by rerdavies** [comment]
The default lv2 path is 

    /usr/lib/lv2:/usr/local/lib/lv2:/usr/modep/lv2

If that isn't sufficient, you an edit the setting lv2_path in

/etc/pipedal/config/config.json:

      "lv2_path": "/usr/lib/lv2:/usr/local/lib/lv2:/usr/modep/lv2",

with the caveat that I believe it gets overwritten each time you update PiPedal.

I should probably set a up system whereby you can create persistent overrides in '/var/pipedal/config' as well.

**Reply by BorisSutin** [reply]
Thank you, I already found the path to the config in the source code yesterday. I liked your project, looked at the code, interesting implementation. Works faster than modep, although there are still problems and bugs. For example, if the preset contains many nodes and plugins, it takes a very long time to load. Yesterday I played for about 30 minutes, after which the sound began to disappear. Doing a reset on the audio settings page, applying new buffer parameters, it appeared for a short time, then disappeared again. And I also wanted to ask if you have plans to implement scenes or snapshots, this will take the project to a new level. But in any case, thank you very much for your work, do not abandon the project, good luck!)

**Reply by rerdavies** [reply]
I'm not familiar with the concept. What are scenes and snapshots? I can guess what scenes might be, but no idea what a snapshot would be.

> sound disappears

Are you working on the Pi when it happens? I've notice bookworm seems less intolerant of eternal load on the pi than buster did, but I haven't worked out the pattern yet.  Could be that I'm using an external USB ssd drive instead of sd-cards these days.

**Reply by BorisSutin** [reply]
Scenes or snapshots are an established concept in all modern guitar processors. The concept is very simple. One song, one preset, but it has several states. For example, intro, verse, chorus. The guitar can sound different everywhere, somewhere more reverb, somewhere delay, clean or crunch. Changing presets takes a long time, but changing snapshots or scenes is instant and this is very convenient. By the way, there are snapshots in modep too.

As for the sound problems, I will try to help you somehow. I will conduct more thorough testing. I have a pi5 with the adau1761 codec, USB flash 3.0. Buffer parameters are 64 samples and 2 periods.

**Reply by rerdavies** [reply]
You might want to try using 64x3 or 32x3 buffers. USB audio adapters (if your device is a USB audio adapter) vastly prefer x3 buffers, probably because USB audio packets contain 48 samples). If it's an internal device, the same may be true too.

If you have your USB flash drive connected to a USB 3.0 port, you should ensure that your USB adapter is connected to a USB 2.0 port so that they don't share the same USB bus. USB 2.0 provides more than adequate bandwidth for audio. 

Scenes: consider it officially added as a high-ish priority feature request.

**Reply by rerdavies** [reply]
@BorisSutin 

So I've been doing some thinking on scenes/snapshots. And I'd be curious as to your thoughts on the following...

I'm not sure I like either Helix's instances, or MOD's scene management scheme, for various reasons. Helix because instances are a deep feature that is difficult to surface in the UI (how do you SEE what instance you're using, and what the next or previous instance is, and how do you jump to to an instance from the UI, and why do you only get four, and isn't it only a really useful feature if you have spent hours progamming your footswitch controller to do it?);  MOD because it seems like they've provided a bad implementation of actual scene management in the sense that a stage manager would expect in order to cover up the "don't reload the plugins" problem. 

It seems to me that the fundamental issue that both implementations are trying to address is not reloading plugins in order to avoid glitching audio output when switching presets.

A much simpler solution (I think) is  for PiPedal to implement the following feature: If you load a new preset, and the structure is identical to the structure of the previous preset, don't reload the LV2 plugins. That allows you to use banks for "scene management", and "snapshots" are pretty much automatic. 

I have this feature in my backlog, but the original conception was to recycle plugin instances based on a fairly complicated version of "similar structure", stealing of all or maybe just some plugin instances from the previous running preset.  A risky feature that's complicated to implement.

But if it's only done for presets with identical structure, the feature is pretty easy to implement, and I think does what you're asking for. And I think all the necessary de-popping and zipping is done already for route splits, and bypasses (smooth fades between routes on splits, and smooth blended fades between processed and unprocessed audio on bypass).

An additional feature might be a "scene management" dialog which just shows all the presets in the current bank, and has big next/previous buttons and click-selection, in order to allow easy on-stage switching of presets during performance. Footswitch implemetnation is already provided by MIDI bindings. Not sure what I would call it, though.  It seems like a useful feature, but "Scene Management" doesn't seem like the right name.

I don't have direct experience with any of the solutions you're talking about. Perhaps I'm missing something. So you might be able to provide insight that I'm missing.

What do you think?

**Comment by BorisSutin** [comment]
Believe me, everything you wrote has nothing to do with the sound problem. Yesterday's test showed that the sound was lost due to the use of third-party lv2 NAM and IR players. Profiles are not loaded correctly and the sound is lost. That is, I select some profile, then go to another plugin block, select what is there, come back and see that the profile field is empty again. Now I use your plugins TooB

Another small question, I still don't understand, do you use Jack for connections or everything is created only with Alsa

**Comment by rerdavies** [comment]
I'm using strictly ALSA. Jack support for services seems to have been
broken in bookworm, and pipewire doesn't seem to actually be up to low
latency audio processing. Plus no contention with pulseaudio.

ALSA seems to solve so many problems. At the expense, of course, of Jack
routing which wouldn't work for pipedal anyway

**Comment by rerdavies** [comment]
If you could let me know which plugins you were using, I'll take a look at
it. The problem of synchronizing state for filename  properties is...
Complicated.

**Reply by BorisSutin** [reply]
NAM Player and in your version of the plugin TooB this also happens. When adding to the chain, it sometimes gives an error message.  And how to enable logging, is there such a possibility?

**Reply by rerdavies** [reply]
It would be nice to know what the error message is. If you see it again, reply back and let me know.

Plugin log message show up in the systemd log files.  

     journalctl -b 0 | grep pipedald

although they tend to be rather sparse.

There is an important fix for TooB Convolution Reverb and Cab IR plugins. Watch out for it. The fix has, unfortunately been tied up by a larger feature. But it should arrive sometime today.

I will install NAM player to see what it's doing.

**Reply by rerdavies** [reply]
@BorisSutin 

I can't seem to find any LV2 plugin named "NAM Player".  You didn't, by any chance, mean "Neural Amp Modeller"? 

Could you provide me with a link to where you got it, please. (Fine if it's a modep plugin).

**Reply by BorisSutin** [reply]
I meant  NAM and  IR Player ) Sorry for not writing the exact names. URI: http://github.com/mikeoliphant/neural-amp-modeler-lv2 URI: https://mod.audio/plugins/CabinetLoader  Their use is not essential for me, I just wanted to try third-party lv2 and see how they will work with your project. I took the most frequently used ones in modep

**Reply by BorisSutin** [reply]
https://drive.google.com/file/d/1l7Zih_0QEvi8BvLb33ETCJ6LTIns4LNs/view?usp=sharing
You asked me to write about the error if I see it again. I made a screen capture to make it clearer.

**Reply by rerdavies** [reply]
@BorisSutin 

FYI, the procedure for overriding the LV2 Plugin path has changed in v1.2.44. 

In previous versions, the config.json file was overwritten by installer upgrades. Instead of customizing /etc/pipedal/config/config.json, make your changes to /var/pipedal/config/config.json (where the changes will persist after PiPedal is upgraded). Once you open /var/pipedal/config/config.json for editing, I'll think you'll get the gist of what happened.

**Reply by rerdavies** [reply]
@BorisSutin:

I built and installed Mike Oliphant's Neural Amp modeler. It actually seems to work without problems under version 1.2.44 of PiPedal, including updating of the filenames in the UI.

There was a fix that could have quite possibly interfered with updating of File controls in prior versions.

If you get a chance, could you try again with PiPedal  v1.2.44 (released today.

**Reply by rerdavies** [reply]
@BorisSutin 

OK. I reproduced the issue on Toob Neural Amp Modeller. Fixes will be included in the next release.

I did try Mike Oliphant's version of Neural Amp Modeller. It works fine on PiPedal 1.2.44, although there is a fix in 1.2.44 for a bug that might have affected it. 

If you're using the MOD version of Neural Amp modeller, and the model does not initially display, and does not load or save properly, that would be a bug with the MOD plugin. Mike Oliphant and I worked to implement fixes for those issues in his code. The MOD plugin derives from Mike's version. It's possible they haven't merged fixes for those issues. If that's the case, please file a bug with the MOD project.

**Comment by rerdavies** [comment]
Thanks. I'll give it a try.

**Comment by BorisSutin** [comment]
I won't brag, but I have a lot of experience with guitar equipment. I've tried everything that's on the market, Kemper, QuadCortex, Fractal, budget segment. I owned a Fractal for several years and shot tutorials on YouTube because the device is very complicated in terms of settings, but it provides huge opportunities for musicians and is very flexible in terms of settings. You can watch my channel, if you want to know more about me. https://www.youtube.com/c/SkyHawkGarage

Therefore, the request to add (scenes/state) is not my desire, the fact is that I know the needs of musicians very well. And I know how to take your project to a new level. I still have many suggestions, advice and comments. We will implement everything step by step. I am a programmer by education and a guitarist at heart)

Let's go over the scenes again. You can use any name for this function, they are different in all processors. You are right that the main problem when switching presets is the time to create a host for the plugin and its audioports, nodes. In long chains, switching can take from 3 to 5-7 seconds. Switching between states in the preset itself happens instantly. Or here is a simple example, I have a looper in the chain, I recorded the first loop and want to add bass guitar emulation, I turn on the plugin with a lowering of the tuning (pitch shifter), I record another overdub loop, then I want to play a solo, and I need a new setting in the preset again. In terms of using scenes, this is just one button press. This is impossible to do with switching a preset, at least because the looper will lose everything that was in its memory. The only way is to put the looper in static blocks and always keep it stationary at the end of the chain. But this is completely rational. This is just a small example. Switching between presets during a song by scrolling through them is extremely inconvenient. As I wrote earlier, one preset is one song, and it has many states. There are musicians who have 500-1000 songs, especially cover bands.


https://github.com/user-attachments/assets/a09b5a8a-c54f-4f84-bcbf-1a9b7747b915



Now to the issue of control. Look at the example of how I implemented it for Modep. I will attach a video to you. I wrote a shell on Flutter, which runs on Raspberry without a desktop. I plan to write the same for managing your project and will definitely give you the finished assembly later, place it in your repository so that people can use it. In my project, there will be three foot controllers connected to the Raspberry Pi gpio. For use with a midi controller, I suggest using the parameter of the programs change PC from 1 to 8. I have never seen more than eight scenes anywhere, 8 scenes are more than enough. In your editor interface, I would make a column with a list of scenes on the left or right. Or a drop-down combo box like for presets. Add a state array to the preset structure that specifies all the parameters of all the plugins of the current preset, including paths to profiles. Add a method to the websocket to getCurrentState of the preset, a method for getAllState of available scenes. In the preset, add the default scene parameter, the one that will be immediately activated when the preset is loaded.

I hope Google Translate translated my thoughts correctly)

Yes, I completely forgot to tell you, I still couldn’t update to version 44, in About it it still shows 42.experimental. I tried to delete the pipedal package and reinstall it. It didn’t help. Check the package itself, maybe something was packed incorrectly.
![pipedal_vers](https://github.com/user-attachments/assets/1addeec7-3136-495c-8406-df0c918e7592)


And also about using banks. Recently, on top guitar processors, this concept has been replaced by setlists. For example, I come to play with a band at a performance, I choose a pre-planned list of songs/presets and see only them in the list, it’s convenient for me to switch between them in order, I don’t have to scroll through a bunch of presets to find the right one. Thus, the setlist contains links to presets necessary for a specific performance program.Although in essence, both banks and setlists have almost the same function This is very convenient, last year, at the request of musicians, I wrote a separate mobile application for Android and iOS to control the Fractal with the setlist function. But setlists are later, at this stage I would like at least the implementation (scenes / states) I understand that it will be necessary to write a lot of code, but we are in no hurry, we just want to make the world of musicians better. After all, top processors are very expensive, not everyone can afford to buy them. And projects such as pipedal and modep can very well get someone more for less money. 

God bless you! Good luck to you!)
